examples/Efficient Cavity Control with SNAP Gates.py

```diff
@@ -0,0 +1,231 @@
+
+theta1, theta2, theta3 = pad_thetas(N, theta1), pad_thetas(N, theta2), pad_thetas(N, theta3)
+
+initial = basis(N, 0) # initial vacuum state
+target = (jnp.sqrt(3) * basis(10, 3) +  basis(10, 9)) / 2.0 # target state b1 as shown above|
+
+
+# # Gradient Descent -- Case in point for `qgrad`
+# 
+# This is where the power of `qgrad` comes in. Since qgrad's functions used in this notebook -- basis, to_dm, dag, Displace, fidelity -- support JAX, we can evaluate the gradient of the cost function in one line using JAX's `grad`. This saves us painstakingly evaluating the derivative of the cost function analytically.
+# 
+
+# In[6]:
+
+
+epochs = 17
+lr = 0.1 # learning rate
+max_iters = 100
+iters = 0
+params = jnp.concatenate((alphas, theta1, theta2, theta3)).reshape(3 * N + 3, 1)
+der_cost = grad(cost) # autodiff of the cost function
+
+params_hist = [params]
+
+for epoch in range (epochs):
+    iters = 0
+    while iters  < max_iters:
+        prev_params = params
+        der = der_cost(prev_params, initial, target, T, N)
+        params = prev_params - lr * der 
+        iters += 1
+    cost_val = cost(params, initial, target, T, N)
+    progress = [epoch+1, cost_val]
+    if ((epoch + 2) % 2 != 0):
+        params_hist.append(params)
+        print("Epoch: {:2f} | Cost: {:3f}".format(*jnp.asarray(progress)))
+
+
+# In[13]:
+
+
+def display_evolution(parameters):
+    """
+    Displays the Wigner function of a state during the learning schedule
+    """
+    fig, ax = plt.subplots(6, 2, figsize=(10, 20))
+    vacuum = onp.array(jnp.dot(initial, dag(initial)))
+    hinton(Qobj(vacuum), ax=ax[0][0])
+    plot_wigner(Qobj(vacuum), ax=ax[0][1])
+    for i in range(4):
+        pred = add_blocks(initial, T, N, parameters[2 * i])
+        dm = jnp.dot(pred, dag(pred)) # This converting back and
+        dm = onp.array(dm)            # forth between np and jnp is why we need qgrad
+        hinton(dm, ax = ax[i+1][0])
+        plot_wigner(Qobj(dm), ax=ax[i+1][1])
+    
+    # final parameters after 17th epoch
+    pred =  add_blocks(initial, T, N, params)
+    dm = onp.array(jnp.dot(pred, dag(pred))) 
+    hinton(dm, ax = ax[5][0])
+    plot_wigner(Qobj(dm), ax=ax[5][1])
+    
+    plt.subplots_adjust(hspace = 0.5, wspace = 0.5)
+    plt.show()
+    
+display_evolution(params_hist)
+
+
+# # Comments on Learning
+# We see that starting from a vacuum state $|0>$, we efficiently learn the target state
+# $b_{1} = \frac{\sqrt3 |3> + |9>}{2}$. The desired target state's Hinton plot and Wigner function are shown below. It can be seen that the last row above is almost the same as the target state shown below, implying we learn parameters $\alpha$ and $\vec \theta$ efficiently so as to go from the vacuum state to a desired target state (in this case a binomial state) in just three applications of blocks, $\hat B$. This tutorial is just meant to show the usage of `qgrad`. One may increase the number of epochs to achieve even better learning.
+
+# In[15]:
+
```

Comment:
This could be titled as `Conclusion`

Note that we have not used the smart block initialisation that the Fossel paper uses, mention this in the beginning and also here so that users are not confused since we are not doing everything that the Fossel paper does. 

Also another point to mention would be that we use a very simple gradient update rule here, a simple Euler forward method. This could be replaced for instance with Adam: https://jax.readthedocs.io/en/latest/_modules/jax/experimental/optimizers.html#adam 

This is very easy to implement with Jax: 

Note that for 1000 iterations I got stuck in some local minima with the following loss and final state

```
from jax.experimental import optimizers
from jax import jit


alphas = jnp.array([1., 0.5, 1.]) # Displace parameters

theta1, theta2, theta3 = [0.5], [0.5, 1.5, 0.5], [0.5, 1.5, 0.5, 1.3] # SNAP parameters
# NOTE: No input values to JAX differentiable functions should be int
thetas = jnp.array([pad_thetas(N, p) for p in [theta1, theta2, theta3]])

opt_init, opt_update, opt_get_params = optimizers.adam(step_size=0.01)
opt_state = opt_init((alphas, thetas))

# Define a compiled update step

@jit
def step(i, opt_state, initial, target):
    alphas, thetas = opt_get_params(opt_state)
    g = der_cost(alphas, thetas, initial_state, target_state)
    return opt_update(i, g, opt_state)


loss = []

for i in range(1000):
    opt_state = step(i, opt_state, initial_state, target_state)
    alphas, thetas = opt_get_params(opt_state)
    loss.append(cost(alphas, thetas, initial_state, target_state))

    if i % 200 == 0:        
        state_evolved = apply_blocks(alphas, thetas, initial_state)
        show_state(state_evolved)
        
plt.plot(loss)
```


![Unknown-2](https://user-images.githubusercontent.com/6968324/88790526-9a2b8900-d198-11ea-8cc9-be8f2a28282c.png)


![image](https://user-images.githubusercontent.com/6968324/88790544-9ef03d00-d198-11ea-90c4-c5937955b0d9.png)

Reply:
Note that this might be due to jax optimisers not working with complex numbers or the way we have separated alphas and thetas in the params and not really let them be some matrix or vector.

I do get some warning there for complex numbers. So please try to check that. Alternatively it is extremely straight forward just to replace your gradient descent lines with the following from jax optimisers which implements Adam. Here we will need to initialise our own m0, v0 for the alphas and thetas and update both of them (not just x here which is probably just considered to be the full vector of parameters)

```
  step_size = make_schedule(step_size)
  def init(x0):
    m0 = jnp.zeros_like(x0)
    v0 = jnp.zeros_like(x0)
    return x0, m0, v0
  def update(i, g, state):
    x, m, v = state
    m = (1 - b1) * g + b1 * m  # First  moment estimate.
    v = (1 - b2) * jnp.square(g) + b2 * v  # Second moment estimate.
    mhat = m / (1 - b1 ** (i + 1))  # Bias correction.
    vhat = v / (1 - b2 ** (i + 1))
    x = x - step_size(i) * mhat / (jnp.sqrt(vhat) + eps)
    return x, m, v
  def get_params(state):
    x, m, v = state
    return x
  return init, update, get_params
```